ulib/LowStar.Vector.fst

```diff
@@ -0,0 +1,558 @@
+
+module U32 = FStar.UInt32
+
+/// Some helpers for `LowStar.Buffer`
+
+val get_as_seq_index:
+  #a:Type -> h:HS.mem -> buf:B.buffer a -> i:uint32_t{i < B.len buf} ->
+  Lemma (B.get h buf (U32.v i) == S.index (B.as_seq h (B.gsub buf i 1ul)) 0)
+let get_as_seq_index #a h buf i = ()
+
+val get_preserved_buffer:
+  #a:Type -> buf:B.buffer a ->
+  i:uint32_t -> len:uint32_t{U32.v i + U32.v len <= B.length buf} ->
+  k:uint32_t{(k < i || i + len <= k) && k < B.len buf} ->
+  h0:HS.mem -> h1:HS.mem ->
+  Lemma (requires (B.live h0 buf /\
+		  modifies (B.loc_buffer (B.gsub buf i len)) h0 h1))
+	(ensures (B.get h0 buf (U32.v k) == B.get h1 buf (U32.v k)))
+let get_preserved_buffer #a buf i len k h0 h1 =
+  get_as_seq_index h0 buf k; get_as_seq_index h1 buf k
+
+/// Abstract vector type
+
+noeq type vector_str a =
+| Vec: sz:uint32_t ->
+       cap:uint32_t{cap >= sz} ->
+       vs:B.buffer a{B.len vs = cap} -> 
+       vector_str a
+
+val vector (a: Type0): Tot Type0
+let vector a = vector_str a
+
+/// Specification
+
+val as_seq: 
+  HS.mem -> #a:Type -> vec:vector a -> 
+  GTot (s:S.seq a{S.length s = U32.v (Vec?.sz vec)})
+let as_seq h #a vec =
+  B.as_seq h (B.gsub (Vec?.vs vec) 0ul (Vec?.sz vec))
+
+/// Capacity
+
+unfold val size_of: #a:Type -> vec:vector a -> Tot uint32_t
+unfold let size_of #a vec = 
+  Vec?.sz vec
+
+unfold val capacity_of: #a:Type -> vec:vector a -> Tot uint32_t
+unfold let capacity_of #a vec =
+  Vec?.cap vec
+
+unfold val is_empty: #a:Type -> vec:vector a -> Tot bool
+unfold let is_empty #a vec =
+  size_of vec = 0ul
+
+unfold val is_full: #a:Type -> vstr:vector_str a -> GTot bool
+unfold let is_full #a vstr =
+  Vec?.sz vstr >= max_uint32
+
+/// Memory-related
+
+unfold val live: #a:Type -> HS.mem -> vector a -> GTot Type0
+unfold let live #a h vec =
+  B.live h (Vec?.vs vec)
+
+unfold val freeable: #a:Type -> vector a -> GTot Type0
+unfold let freeable #a vec =
+  B.freeable (Vec?.vs vec)
+
+unfold val loc_vector_within: 
+  #a:Type -> vec:vector a -> 
+  i:uint32_t -> j:uint32_t{i <= j && j <= size_of vec} -> GTot loc
+unfold let loc_vector_within #a vec i j =
+  B.loc_buffer (B.gsub (Vec?.vs vec) i (j - i))
+
+unfold val loc_vector: #a:Type -> vector a -> GTot loc
+unfold let loc_vector #a vec =
+  B.loc_buffer (Vec?.vs vec)
+
+unfold val loc_addr_of_vector: #a:Type -> vector a -> GTot loc
+unfold let loc_addr_of_vector #a vec =
+  B.loc_addr_of_buffer (Vec?.vs vec)
+
+unfold val frameOf: #a:Type -> vector a -> Tot HH.rid
+unfold let frameOf #a vec =
+  B.frameOf (Vec?.vs vec)
+
+unfold val hmap_dom_eq: h0:HS.mem -> h1:HS.mem -> GTot Type0
+unfold let hmap_dom_eq h0 h1 =
+  Set.equal (Map.domain (MHS.get_hmap h0))
+	    (Map.domain (MHS.get_hmap h1))
+
+val modifies_as_seq:
+  #a:Type -> vec:vector a -> dloc:loc ->
+  h0:HS.mem -> h1:HS.mem ->
+  Lemma (requires (live h0 vec /\ 
+		  loc_disjoint (loc_vector vec) dloc /\
+		  modifies dloc h0 h1))
+	(ensures (as_seq h0 vec == as_seq h1 vec))
+	[SMTPat (live h0 vec); 
```

Comment:
This pattern most likely won't work, owing to `live` defined as `unfold`

Reply:
the `unfold`s above are likely causing a bunch of issues, since `loc_vector` is also defined as unfold; Tahina and I recommend removing the `unfold`s above, and leaving `inline_for_extraction` for the small helpers you hope to see disappear when extracting; if there was a deeper reason for using `unfold`, let us know!

(the general rule of thumb is that a definition marked as unfold is inlined before going to SMT, meaning that it will not be able to be matched against)

Reply:
Tahina says it's a good practice to write `live h1 vec` in your post-condition because it's almost always going to be useful and will save z3 some search effort.